### PR TITLE
Analytics Feature Branch

### PR DESCRIPTION
**Pull request by nestorw**
I wanted to get some discussion going on moving the analytics/reporting work into master. 

This is the feature branch for the analytics working group and represents the work done so far in [the analytics project](https://github.com/samvera/hyrax/projects/3). As many new features as possible are behind the new `analytics_redesign` Flipflop feature and shouldn't impact existing projects.
This will require a `rails db:migrate` since a new `resource_stats` table was added.

@samvera/hyrax-code-reviewers

### CODE REVIEW

**Comment by no-reply** [comment]
I'd love to work on getting this into `master` ASAP. I've assigned a self-review and will give this time in the next 2-3 days.

File: app/assets/javascripts/hyrax/dashboard_actions.js [hunk 1]
```diff
@@ -32,4 +32,27 @@ Blacklight.onLoad(function() {
     return false;
   });
 
+  // Create sortable, searchable table
+  $('#analytics-collections-table').DataTable();
```
Comment (jcoyne) [inline app/assets/javascripts/hyrax/dashboard_actions.js]: This code, is not about the dashboard page itself, but about an analytics component that is on the dashbord page. Can you move this code to it's own file that is only about the analytics component of the page?

File: config/initializers/analytics.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+
+  case Hyrax.config.analytics
+  when 'matomo'
+    require 'piwik'
+    Piwik::PIWIK_URL = config['matomo_url']
+    Piwik::PIWIK_TOKEN = config['matomo_token']
+
+    Hyrax::Analytics::Matomo.config = config
+    Rails.logger.error("Invalid Matomo config from #{filename}.") unless Hyrax::Analytics::Matomo.valid?
+  when 'google' || true
```
Comment (jcoyne) [inline config/initializers/analytics.rb]: Should this be `else`?

File: config/routes.rb [hunk 3]
```diff
@@ -109,6 +109,10 @@
   resources :dashboard, only: [] do
     collection do
       get 'activity', action: :activity, as: :dashboard_activity
+      get 'update_collections', action: :update_collections, as: :update_collections
```
Comment (jcoyne) [inline config/routes.rb]: These actions should be on their own controller (or maybe 3 separate controllers) specifically for analytics.

File: app/controllers/hyrax/admin/reports/works_controller.rb [hunk 4]
```diff
@@ -0,0 +1,11 @@
+module Hyrax
+  class Admin::Reports::WorksController < ApplicationController
+    with_themed_layout 'dashboard'
+
+    def status; end
```
Comment (jcoyne) [inline app/controllers/hyrax/admin/reports/works_controller.rb]: I'm not seeing any authorization checks.  Is that expected?

File: app/presenters/hyrax/menu_presenter.rb [hunk 5]
```diff
@@ -42,6 +42,18 @@ def user_activity_section?
       end
     end
 
+    # @return [Boolean] true if the current controller happens to be one of the controllers that deals
+    # with admin reports. This is used to keep the parent section on the sidebar open
+    def admin_reports_section?
+      %w[works].include?(controller_name) && %w[status attributes activity].include?(action_name)
```
Comment (jcoyne) [inline app/presenters/hyrax/menu_presenter.rb]: Why `%w[works].include?(controller_name)` why not `'works' == controller_name`?

File: app/services/hyrax/works_count_service.rb [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+      works.map do |work|
+        created_date = DateTime.parse(work['system_create_dtsi']).in_time_zone.strftime("%Y-%m-%d")
+        SearchResultForWorkCount.new(work, created_date, 0, work['human_readable_type_tesim'][0], work['visibility_ssi'])
+      end
+    end
+
+    private
+
+      def builder(_)
+        search_builder.new(context).rows(1000)
```
Comment (jcoyne) [inline app/services/hyrax/works_count_service.rb]: Create a `MAX_ROWS = 1000` constant for this.

File: app/services/hyrax/count_service.rb [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+    def search_results(access)
+      response = context.repository.search(builder(access))
+      response.documents
+    end
+
+    protected
+
+      # @param [Symbol] access :read or :edit
+      def builder(access)
+        search_builder.new(context, access, @model).rows(100)
```
Comment (jcoyne) [inline app/services/hyrax/count_service.rb]: Create a `MAX_ROWS = 100` constant for this.

File: app/views/_analytics.html.erb [hunk 8]
```diff
@@ -0,0 +1,6 @@
+<% case Hyrax.config.analytics %>
+<%when 'matomo' %>
+  <%= render partial: '/matomo', formats: [:html] %>
+<%when 'google' || true%>
```
[the remark this answers is not in the document]
Comment (jcoyne) [reply app/views/_analytics.html.erb]: I see what you're trying to do.  Shouldn't it be `'google', true`?

File: config/initializers/analytics.rb (hunk 2 above)
Comment (jcoyne) [reply config/initializers/analytics.rb]: I see, I think `'google', true` is what you want.

File: app/controllers/hyrax/dashboard_controller.rb [hunk 9]
```diff
@@ -9,12 +9,34 @@ def show
...
+        @collection_rows = Hyrax::CollectionsCountService.new(self, Hyrax::AdminSetSearchBuilder, ::Collection).search_results_with_work_count(:read)
+        @work_rows = Hyrax::WorksCountService.new(self, Hyrax::WorksSearchBuilder).search_results_with_work_count(:read)
         render 'show_admin'
       else
         @presenter = Dashboard::UserPresenter.new(current_user, view_context, params[:since])
         render 'show_user'
       end
     end
+
+    def update_collections
```
[the remark this answers is not in the document]
Comment (lfarrell) [reply app/controllers/hyrax/dashboard_controller.rb]: That method goes away in a later branch I'm working on.

File: app/presenters/hyrax/admin/repository_growth_presenter.rb [hunk 10]
```diff
@@ -1,15 +1,22 @@
 module Hyrax
   module Admin
     class RepositoryGrowthPresenter
-      def initialize
-        @x_min = 90.days.ago.beginning_of_day
+      def initialize(time_period = 90)
+        @x_min = Integer(time_period).days.ago.beginning_of_day
```
[the remark this answers is not in the document]
Comment (lfarrell) [reply app/presenters/hyrax/admin/repository_growth_presenter.rb]: It should always be an integer

File: app/services/hyrax/admin_set_service.rb [hunk 11]
```diff
@@ -2,23 +2,7 @@
...
-      @search_builder = search_builder
-    end
-
-    # @param [Symbol] access :deposit, :read or :edit
-    def search_results(access)
-      response = context.repository.search(builder(access))
-      response.documents
-    end
-
+  class AdminSetService < CountService
```
[the remark this answers is not in the document]
Comment (lfarrell) [reply app/services/hyrax/admin_set_service.rb]: Yes. We've been leery of making breaking changes.

**Pull request closed without merging**